ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/AbstractTransactionsLayer.java

```diff
@@ -0,0 +1,593 @@
+  }
+
+  protected abstract void internalNotifyAdded(
+      final NavigableMap<Long, PendingTransaction> senderTxs,
+      final PendingTransaction pendingTransaction);
+
+  @Override
+  public PendingTransaction promote(final Address sender, final long nonce) {
+    final var senderTxs = txsBySender.get(sender);
+    if (senderTxs != null) {
+      long expectedNonce = nonce + 1;
+      if (senderTxs.firstKey() == expectedNonce) {
+        final PendingTransaction promotedTx = senderTxs.pollFirstEntry().getValue();
+        processRemove(senderTxs, promotedTx.getTransaction(), PROMOTED);
+        metrics.incrementRemoved(promotedTx.isReceivedFromLocalSource(), "promoted", name());
+
+        if (senderTxs.isEmpty()) {
+          txsBySender.remove(sender);
+        }
+        return promotedTx;
+      }
+    }
+    return nextLayer.promote(sender, nonce);
+  }
+
+  private TransactionAddedResult addToNextLayer(
+      final PendingTransaction pendingTransaction, final int distance) {
+    return addToNextLayer(
+        txsBySender.getOrDefault(pendingTransaction.getSender(), EMPTY_SENDER_TXS),
+        pendingTransaction,
+        distance);
+  }
+
+  private TransactionAddedResult addToNextLayer(
+      final NavigableMap<Long, PendingTransaction> senderTxs,
+      final PendingTransaction pendingTransaction,
+      final int distance) {
+    final int nextLayerDistance;
+    if (senderTxs.isEmpty()) {
+      nextLayerDistance = distance;
+    } else {
+      nextLayerDistance = (int) (pendingTransaction.getNonce() - (senderTxs.lastKey() + 1));
+    }
+    return nextLayer.add(pendingTransaction, nextLayerDistance);
+  }
+
+  private void processAdded(final PendingTransaction addedTx) {
+    pendingTransactions.put(addedTx.getHash(), addedTx);
+    final var senderTxs = txsBySender.computeIfAbsent(addedTx.getSender(), s -> new TreeMap<>());
+    senderTxs.put(addedTx.getNonce(), addedTx);
+    increaseSpaceUsed(addedTx);
+    metrics.incrementAdded(addedTx.isReceivedFromLocalSource(), name());
+    internalAdd(senderTxs, addedTx);
+  }
+
+  protected abstract void internalAdd(
+      final NavigableMap<Long, PendingTransaction> senderTxs, final PendingTransaction addedTx);
+
+  protected abstract int maxTransactionsNumber();
+
+  private void evict(final long spaceToFree, final int txsToEvict) {
+    final var evictableTx = getEvictable();
+    if (evictableTx != null) {
+      final var lessReadySender = evictableTx.getSender();
+      final var lessReadySenderTxs = txsBySender.get(lessReadySender);
+
+      long evictedSize = 0;
+      int evictedCount = 0;
+      PendingTransaction lastTx;
+      // lastTx must never be null, because the sender have at least the lessReadyTx
+      while ((evictedSize < spaceToFree || txsToEvict > evictedCount)
+          && !lessReadySenderTxs.isEmpty()) {
+        lastTx = lessReadySenderTxs.pollLastEntry().getValue();
+        processEvict(lessReadySenderTxs, lastTx);
```

Comment:
Following this LayersTest scenario...
```
            new Scenario("overflow to sparse")
                .addForSender(S1, 0, 1, 2, 3, 4, 5, 6)
                .expectedPrioritizedForSender(S1, 0, 1, 2)
                .expectedReadyForSender(S1, 3, 4, 5)
                .expectedSparseForSender(S1, 6)),
```

Why do we get an `evictableTx` of nonce = 3 (via `orderByMaxFee.first()`) and then later switch it out to be `lastTx` where nonce = 6?
Maybe it's just naming, but I was expecting `evictableTx` to be nonce = 6 straight away.

Reply:
The rule when evicting is to avoid to introduce gaps, so we get the least valuable tx, then we get the sender of it, and we evict the tx with the highest nonce for that sender.
Could a better naming be `evictableSenderTx`? other suggestions?